Add method `has_control_flow_op()` to `QuantumCircuit`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
This method was added to fix a performance regression introduced by #13921. This check should be quicker than the multiple calls to `isinstance()`.


### Details and comments

This new method for `QuantumCircuit` performs a fast path to determine if a circuit has a control flow operation present within. This is performed by using `QuantumCircuit.count_ops()` and checking if any of the names there belong to a `ControlFlowOp`.

Additionally, this method fixes an unintentional performance regression caused by #13921.

raynelfss committed:
Add: Add method has_control_flow_op(), to `QuantumCircuit`

This method was added to fix a performance regression introduced by #13921. This check should be quicker than the multiple calls to `isinstance`.

jakelishman commented:
I'm not sure we need to do this beyond the feature-freeze deadline. There's already no need to do all the `isinstance` checks - the original PR could have written
```python
any(instruction.is_control_flow() for instruction in circuit.data)
```

raynelfss committed:
Lint: Fix formatting

raynelfss committed:
Lint: Removed unused import and f-string

raynelfss committed:
Docs: Add release note
- Add `has_control_flow_op` to autogenerate docs.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -423,6 +423,15 @@ impl CircuitData {
+        let operations = self.count_ops();
+        CONTROL_FLOW_OP_NAMES
+            .iter()
+            .any(|name| operations.contains_key(name))
```
`count_ops()` will iterate over the circuit anyway. It'll be better to just do:

```suggestion
        self.data.iter().any(|x| x.control_flow())
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -70,7 +70,8 @@ use std::f64::consts::PI;
 use std::sync::OnceLock;
 
-static CONTROL_FLOW_OP_NAMES: [&str; 4] = ["for_loop", "while_loop", "if_else", "switch_case"];
+pub(crate) static CONTROL_FLOW_OP_NAMES: [&str; 4] =
```
With the above comment we don't need this change.

```suggestion
static CONTROL_FLOW_OP_NAMES: [&str; 4] =
```

mtreinish commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -4341,6 +4344,11 @@ def map_calibration(qubits, parameters, schedule):
         )
         return None if inplace else target
 
+    def has_control_flow_op(self) -> bool:
```
@jakelishman's earlier comment is on point here this is past feature freeze so we shouldn't make this public if we want to add it for 2.0. There is a path to fix the regression in python as he suggested, but with my other comments here doing the iteration in rust is faster. We could also still iterate in rust if we did the set intersection with `count_ops()` python side (like is currently in the rust code).

mtreinish reviewed (commented):
Thanks for tackling this. I left a few comments inline. The biggest issue @jakelishman already pointed out that this shouldn't be a new public method for 2.0 so close to the release. There are paths to fixing the performance issue without adding a new method, we can add the public api in 2.1 though.

mtreinish commented:
In this interest of getting a fix for the regression included in 2.0 I pushed up: https://github.com/Qiskit/qiskit/pull/13942 and moved this to 2.1. We should definitely do this after 2.0 is released.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into control_circ_inst

raynelfss committed:
Fix: Addressing review comments

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Docs: Fix formatting issue in `QuantumCircuit` docstring.

raynelfss committed:
Lint: Fix formatting bug in `QuantumCircuit`.

mtreinish reviewed (commented):
Overall this LGTM now, but could you add some tests to test this method explicitly?

raynelfss committed:
Test: Add testcases
- One test case with a regular control flow operation and another with a custom one.

mtreinish commented on test/python/circuit/test_circuit_operations.py:
```diff
@@ -1550,6 +1551,82 @@ def instructions():
+                )
+                # We do not need to validate this custom condition
+                self.condition = condition
+                self._params = [true_body, false_body]
```
This is not supported. We don't allow for custom control flow operations to be defined like this. So we shouldn't have a test for this.

raynelfss replied:
Addressed in 98c2e2c

mtreinish commented on test/python/circuit/test_circuit_operations.py:
```diff
@@ -1550,6 +1551,82 @@ def instructions():
         self.assertEqual(circuit, expected)
         self.assertEqual(circuit.name, "test")
 
+    def test_circuit_has_control_flow_op(self):
```
I would expand this to cover a circuit with each control flow op type and make sure it's detected. The set of defined operations is here: https://github.com/Qiskit/qiskit/blob/main/qiskit/circuit/controlflow/__init__.py#L28

raynelfss replied:
Addressed in 98c2e2c

raynelfss committed:
Test: Remove unsupported test and extend original test.
- Implement a combinatorics based test for control flow operations.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into control_circ_inst

raynelfss committed:
Fix: Import order in `test_circuit_operations`

raynelfss committed:
Merge branch 'main' into control_circ_inst

mtreinish commented on test/python/circuit/test_circuit_operations.py:
```diff
@@ -1576,6 +1581,55 @@ def instructions():
+            circ.cx(0, 1)
+            circ.measure(0, 0)
+            circ.measure(1, 1)
+            self.assertTrue(not circ.has_control_flow_op())
```
```suggestion
            self.assertFalse(circ.has_control_flow_op())
```

mtreinish committed:
Update test/python/circuit/test_circuit_operations.py
